### PR TITLE
Add automatic locking / unlocking of accounts

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**
* Change account signing API to two sign functions;
* Sign without passphrase - works if account is unlocked
* Sign with passphrase - always works and unlocks the account
* Account stays unlocked for X ms and is then automatically locked

### CODE REVIEW

File: accounts/account_manager.go [hunk 1]
```diff
@@ -34,24 +34,33 @@ package accounts
...
-// TODO: pass through passphrase for APIs which require access to private key?
-func NewAccountManager(keyStore crypto.KeyStore2) AccountManager {
+func NewAccountManager(keyStore crypto.KeyStore2, unlockMilliSeconds int) AccountManager {
+	keysMap := make(map[string]crypto.Key)
 	am := &AccountManager{
-		keyStore: keyStore,
+		keyStore:             keyStore,
+		unlockedKeys:         keysMap,
+		unlockedMilliSeconds: unlockMilliSeconds,
+		mutex:                sync.Mutex{}, // for accessing unlockedKeys map
```
Comment (obscuren) [inline accounts/account_manager.go]: No need to initialise a non-pointer.

File: accounts/account_manager.go [hunk 2]
```diff
@@ -34,24 +34,33 @@ package accounts
...
 type Account struct {
 	Address []byte
 }
 
 type AccountManager struct {
-	keyStore crypto.KeyStore2
+	keyStore             crypto.KeyStore2
+	unlockedKeys         map[string]crypto.Key
+	unlockedMilliSeconds int
+	mutex                sync.Mutex
```
Comment (obscuren) [inline accounts/account_manager.go]: Turn this in to a `RWMutex`

File: accounts/account_manager.go [hunk 3]
```diff
@@ -60,11 +69,26 @@ func (am AccountManager) DeleteAccount(address []byte, auth string) error {
 	return am.keyStore.DeleteKey(address, auth)
 }
 
-func (am *AccountManager) Sign(fromAccount *Account, keyAuth string, toSign []byte) (signature []byte, err error) {
+func (am *AccountManager) Sign(fromAccount *Account, toSign []byte) (signature []byte, err error) {
+	am.mutex.Lock()
```
Comment (obscuren) [inline accounts/account_manager.go]: `RLock` / `RUnlock`

File: accounts/account_manager.go [hunk 4]
```diff
@@ -60,11 +69,26 @@ func (am AccountManager) DeleteAccount(address []byte, auth string) error {
...
+	signature, err = crypto.Sign(toSign, unlockedKey.PrivateKey)
+	return signature, err
+}
+
+func (am *AccountManager) SignLocked(fromAccount *Account, keyAuth string, toSign []byte) (signature []byte, err error) {
 	key, err := am.keyStore.GetKey(fromAccount.Address, keyAuth)
 	if err != nil {
 		return nil, err
 	}
+	am.mutex.Lock()
```
Comment (obscuren) [inline accounts/account_manager.go]: Same

File: accounts/account_manager.go [hunk 5]
```diff
@@ -97,3 +119,11 @@ func (am *AccountManager) Accounts() ([]Account, error) {
 	}
 	return accounts, err
 }
+
+func unlockLater(am *AccountManager, addr []byte) {
+	time.Sleep(time.Millisecond * time.Duration(am.unlockedMilliSeconds))
```
Comment (obscuren) [inline accounts/account_manager.go]: Suggestion

```go
select {
    case <-time.After(time.Millisecond * am.unlockMilliseconds):
}
```

File: accounts/account_manager.go [hunk 6]
```diff
@@ -34,24 +34,33 @@ package accounts
...
 // TODO: better name for this struct?
 type Account struct {
 	Address []byte
 }
 
 type AccountManager struct {
-	keyStore crypto.KeyStore2
+	keyStore             crypto.KeyStore2
+	unlockedKeys         map[string]crypto.Key
+	unlockedMilliSeconds int
```
Comment (obscuren) [inline accounts/account_manager.go]: Turn in to a `time.Duration`

File: accounts/accounts_test.go [hunk 7]
```diff
@@ -6,20 +6,69 @@ import (
...
 	a1, err := am.NewAccount(pass)
 	toSign := randentropy.GetEntropyCSPRNG(32)
-	_, err = am.Sign(a1, pass, toSign)
+	_, err = am.SignLocked(a1, pass, toSign)
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	// Cleanup
+	time.Sleep(time.Millisecond * time.Duration(150)) // wait for locking
```
Comment (obscuren) [inline accounts/accounts_test.go]: No need for casting `150 * time.Millisecond`

File: accounts/account_manager.go (hunk 5 above)
Comment (fjl) [inline accounts/account_manager.go]: Why is that better?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Suggestion

select {
    case <-time.After(time.Millisecond * am.unlockMilliseconds):
}‎

Comment (obscuren) [inline accounts/account_manager.go]: because it's going to allow you to extend the locking period.

**Pull request merged**